File: backend/core/rbac.py
from fastapi import Depends, HTTPException

from ..models import User
from .security import current_user


def require_role(*roles: str, detail: str = "无权访问该资源"):
    def _dependency(user: User = Depends(current_user)) -> User:
        if user.role not in roles:
            raise HTTPException(status_code=403, detail=detail)
        return user
    return _dependency


def assert_enterprise_scope(user: User, enterprise_id: int, detail: str = "无权访问该单位数据") -> None:
    if user.role == "enterprise" and user.enterprise_id != enterprise_id:
        raise HTTPException(status_code=403, detail=detail)


def require_operator_manager(user: User = Depends(current_user)) -> User:
    if user.role == "admin":
        return user
    if user.role == "enterprise" and user.is_owner:
        return user
    raise HTTPException(status_code=403, detail="仅单位主管可管理操作员" if user.role == "enterprise" else "无权管理操作员")

File: java-backend/src/main/java/com/xbb/baojing/common/Rbac.java
package com.xbb.baojing.common;

import java.util.Arrays;

/** Ports backend/core/rbac.py — helpers called inline from services/controllers,
 * not annotations, matching the Python codebase's inline-check style exactly
 * (some endpoints check role directly instead of via a dependency, on purpose —
 * see PlanController.addPlan for the one place this distinction actually matters
 * for a test). */
public final class Rbac {
    private Rbac() {}

    public static void requireRole(User user, String detail, String... roles) {
        if (Arrays.stream(roles).noneMatch(r -> r.equals(user.getRole()))) {
            throw ApiException.forbidden(detail);
        }
    }

    public static void assertEnterpriseScope(User user, Integer enterpriseId) {
        assertEnterpriseScope(user, enterpriseId, "无权访问该单位数据");
    }

    public static void assertEnterpriseScope(User user, Integer enterpriseId, String detail) {
        if ("enterprise".equals(user.getRole()) && !user.getEnterpriseId().equals(enterpriseId)) {
            throw ApiException.forbidden(detail);
        }
    }

    public static void requireOperatorManager(User user) {
        if ("admin".equals(user.getRole())) return;
        if ("enterprise".equals(user.getRole()) && user.isOwner()) return;
        throw ApiException.forbidden("enterprise".equals(user.getRole()) ? "仅单位主管可管理操作员" : "无权管理操作员");
    }
}
